addon/test-helper/deprecation.js

```diff
@@ -1,53 +1,50 @@
-/* globals QUnit */
-
-import MethodCallExpectation from "./method-call-expectation";
-
-var NONE = function(){};
+import MethodCallExpectation from './method-call-expectation';
+import MethodCallTracker from './method-call-tracker';
 
 var DeprecationAssert = function(env){
   this.env = env;
-
   this.reset();
 };
 
 DeprecationAssert.prototype = {
-
   reset: function(){
-    this.expecteds = null;
-    this.actuals = null;
+    this.tracker = new MethodCallTracker(this.env.Ember, 'deprecate');
```

Comment:
I'll do a little refactoring to only instantiate a tracker when necessary, but that doesn't address the issue of having to deal with multiple contexts:

```js
expectDeprecation(/foo/);

// tests that could issue deprecation 'foo'

expectDeprecation(function() {
  // test that issues deprecation 'bar'
}, /bar/);

// tests that could also issue deprecation 'foo'
```

AFAICT, `DeprecationAssert` has to deal with the fact that the function given to the second call is not expected to issue the 'foo' deprecation. Essentially there's a single global context for `expectDeprecation` and N number of sandboxed contexts for each call that is passed a function. This is why the `invokeWithNewContext` method creates a new tracker instance and then promptly drops it on the floor.

Although perhaps I've misunderstood this completely o_O